### ISSUE TITLE
zsh: make: function definition file not found

### ISSUE THREAD

**Issue by atwok** [zdharma/zinit#258]
**Describe the bug**
I'm trying to use zinit to manage my fpath.  I found the code below on reddit.  Since the 5.6.2/functions directory contains _* completions as well as a bunch of functions (like zcompinit) I can't just exclude it or my shell immediately breaks.  I've tried the code below with and without as"completions" in it and the failure is the same.

Any _completion that is in the 5.6.2/functions dir doesn't work.  I see the error above for make, git-dir, etc.

I've erased .zcomp* multiple time and then run "zinit compinit" and I swear completions will work occasionally but then they will break on the next shell I open.

(the 4.6.1 directory just holds one completion I'm picking up there - I'm not loading stuff from two version of zsh just to be clear :))

**To Reproduce**
I think this is related:

fpath=( ${fpath[@]:#/usr/local/*functions} )
#fpath=( ${fpath[@]:#/<HOME>/.miniconda3/*functions} )
zinit ice as"completion" wait lucid atclone"print Installing system-completions...; \
              zinit creinstall  /usr/local/modules/4.2.1/init/zsh-functions; \
              zinit creinstall  /<HOME>/.miniconda3/share/zsh/site-functions; \
              zinit creinstall  /<HOME>/.miniconda3/share/zsh/5.6.2/functions" \
        atpull"%atclone" id-as"system-completions"
zinit snippet /dev/null

the rest of my completion related plugins use code that laggardkernel has posted but I can post the exact contents if needed.

**Expected behavior**

the make executable to be run instead of the error in the subject.

**Screenshots**

**Versions:**
 - $ZSH_VERSION == 5.6.2
 - Operating system name and version (uname -a): Centos 6

**Additional context**

**Comment by psprint** [comment]
I've run the following code:

```zsh
fpath=( ${fpath[@]:#$HOME/*polaris/*functions} )
zinit ice as"completion" wait lucid atclone"print Installing system-completions...;
        zinit creinstall /root/.zplugin/polaris/share/zsh/5.7.1-test-2/functions" \
    atpull"%atclone" id-as"system-completions"
zinit snippet /dev/null
```

and my completions are working OK. I'm using a `zsh` that's build with: 

```zsh
zinit id-as"zsh" as"null" notify atclone"./.preconfig; CFLAGS=\"-I/usr/include -I/usr/local/include -g -O2 -Wall\" ./configure --prefix=\"/root/.zplugin/polaris\" &>/dev/null && make install &>/dev/null && rm -f /bin/zsh && cp -vf Src/zsh /bin/zsh" atpull"%atclone" for \
    zsh-users/zsh
```

I.e.: it's functions directory is: `/root/.zplugin/polaris/share/zsh/5.7.1-test-2/functions` and it is being filtered out correctly. Could you provide any more background, e.g.: what do you mean by the `make` function failing – is it the `_make` function actually?

**Comment by atwok** [comment]
I just figured out if I type "zinit csearch" and just hit return, I don't see anything from system-completions.  This seems bad.

I'm just typing "make" at the command line and hitting return and instead of seeing something like "Makefile not found", I get the error above, which I think means zsh is trying to find the _make function and can't?  So the executable isn't even being run.

**Comment by atwok** [comment]
If I try:

fpath=( ${fpath[@]:#/HOME/.miniconda3/share/zsh/*functions} )
zinit ice as"completion" wait lucid atclone"print Installing system-completions...; \
              zinit creinstall  /usr/local/modules/4.2.1/init/zsh-functions; \
              zinit creinstall  /HOME/.miniconda3/share/zsh/5.6.2/functions" \
        atpull"%atclone" id-as"system-completions"
zinit snippet /dev/null

and log out and log in, I see:

zpcompinit: compinit: function definition file not found
Compinit isn't loaded, cannot do compdef replay

Does your functions dir contain _make, _what, _ever and also compinit?  because mine does.  I'm using a compile binary package from conda-forge because I'm using a shared machine and can only change files in my user area.  Not sure if the zsh layout is standard or not.

**Comment by atwok** [comment]
also, when I remove the fpath before using the code above, how does it make it back into the fpath so that when zcompinit is called later it's available?  as I said, that dir does belong in fpath and maybe it needs to be there before the completions get processed later?

just trying to help.  I could be completely off.  :)

**Comment by psprint** [comment]
Yes you're right, you should remove the directory from `fpath` after running `zpcompinit` (now, after the rename, called `zicompinit`). 

Also, it seems that you have some `make` function somewhere in the `fpath`. You can check it via:

```zsh
print $^fpath/make(N)
```

(please paste the output). In general, `creinstall` will install only underscored functions from the directory. If you need some other functions from it then you might need to copy them to some `~/functions` directory and add it to `fpath`:

```zsh
mkdir -p ~/functions
cp $^fpath/[^_]*(N) ~/functions
```

**Comment by atwok** [comment]
Upfront, I know some of this is caused by my installation, but maybe there are some things zinit could do to help me figure out what's going on/may be good features anyway, so I'll put them here but they should probably should be seperate feature requests if you think they are useful.

1)  The code:
```
zinit ice as"completion" atclone"print Installing system-completions...; \
              zinit creinstall  /usr/local/modules/4.2.1/init/zsh-functions; \
              zinit creinstall  /home/USER/.miniconda3/share/zsh/5.6.2/underscores" \
      atpull"%atclone" id-as"system-completions"
zinit snippet /dev/null

```
This creates a snippet called "system-completions" and "dev".  I think the "dev" is probably unintended?

2)  Is there some way to reset everything completion related so that zinit is forced to "rebuild" them all?  Or do you just remove all the plugins/snippets that have completions, logout, then log back in?

I split up my zsh installation.  Everything without an underscore, along with what zsh defines as "utility functions" I left in my functions directory.  Everything else I put in the underscores directory above.

Here's the code you requested:

The first print doesn't return anything.
```
❯ print $^fpath/make(N)

~
09:44:18 ❯ which make
make () {
        # undefined
        builtin autoload -X
}

~
09:44:25 ❯ make
zsh: make: function definition file not found
```

So for what it's worth,  I typed "make -" and then hit tab.  I get up the completion menu:
```
❯ make -
 -- option --
-C  -- change directory first
-I  -- include directory for makefiles
-W  -- pretend file was modified
-f  -- specify makefile
-o  -- specify file not to remake
```

So the completions are working, it's just when I do "make -f foo.txt" instead of getting a fail from make, I get the same function definition file not found.

I do have the definition in my fpath now:
```
❯ echo $FPATH
/usr/local/modules/4.2.1/init/zsh-functions:/home/USER/.zinit/completions:/usr/local/share/zsh/site-functions:/home/USER/.miniconda3/share/zsh/site-functions:/home/USER/.miniconda3/share/zsh/5.6.2/functions:/home/USER/.zinit/snippets/PZT::modules/utility/functions:modules/helper/functions:modules/git/functions:/home/USER/.miniconda3/share/zsh/5.6.2/underscores

❯ ls -l /home/USER/.miniconda3/share/zsh/5.6.2/underscores/_make
-rw-rw-r--. 2 USER GROUP 9528 Nov  7  2018 /home/USER/.miniconda3/share/zsh/5.6.2/underscores/_make
```

Thanks again for all your help.

**Comment by atwok** [comment]
So this worked...?

```
10:00:04 ❯ rm ~/.zcompdump*
rm: remove regular file `/home/USER/.zcompdump'? y
rm: remove write-protected regular file `/home/USER/.zcompdump.zwc'? y

10:01:20 ❯ exec zsh -l
10:01:34 ❯ make
make: *** No targets specified and no makefile found.  Stop.

```
So now make seems to work.  

But if I log out and log back in, make is acting the same way.
Hopefully this helps?

**Comment by psprint** [comment]
Yes this might help. Could you show `grep \'make ~/.zcompdump`?

**Comment by atwok** [comment]
❯ grep \'make ~/.zcompdump
'make' '_make'
'makeinfo' '_texinfo'
'make-kpkg' '_make-kpkg'
'makepkg' '_pkgtool'

**Comment by psprint** [comment]
That is really weird. Could you grep your rcs files for autoload `make`?

**Comment by atwok** [comment]
I grepped around for autoload ```make``` and I didn't find anything.  I did do the following though:

```
❯ grep make .zcompdump
'btmakemetafile' '_bittorrent'
'dmake' '_make'
'gmake' '_make'
'make' '_make'
'makeinfo' '_texinfo'
'make-kpkg' '_make-kpkg'
'makepkg' '_pkgtool'
'pmake' '_make'
'tpkg-make' '_toolchain-source'
            _make _make-kpkg _man _match _math \

~
09:04:08 ❯ rm -rf .zcompdump*

~
❯ exec zsh -l

~
09:04:50 ❯ grep make .zcompdump
'bmake' '_make'
'btmakemetafile' '_bittorrent'
'cmake' '_cmake'
'dmake' '_make'
'freebsd-make' '_make'
'gmake' '_make'
'make' '_make'
'makeinfo' '_texinfo'
'make-kpkg' '_make-kpkg'
'makepkg' '_pkgtool'
'pmake' '_make'
'tpkg-make' '_toolchain-source'
'-value-,CMAKE_GENERATOR,-default-' '_cmake'
            _chsh _cksum _clay _cmake _cmp \
            _madison _mail _mailboxes _main_complete _make \
            _make-kpkg _man _math _math_params _matlab \
            _make _make-kpkg _man _math _math_params \
```

So when I log in, .zcompdump is what is shown in the initial grep.  If I then remove it and do the "exec zsh -l" then .zcompdump is differnt and make works.  Obviously something different is happening in the two cases.  Any idea what that is?

**Comment by psprint** [comment]
> 09:44:18 ❯ which make
> make () {
>         # undefined
>         builtin autoload -X
> }

Above result proves that there has to be an `autoload make` somewhere. The zcomdump grepping didn't explain where it is. You have to look deeper, Also, you could verify that the `make` function is really there by doing `whence -v make` without the `$fpath` entry removed and after running make.

**Comment by atwok** [comment]
I ran zsh with the -x option for the two cases.  Logging in via ssh and then after removing .zcompdump and running "exec zsh -l".  The difference I see is in the "exec" case, compinit actually gets run (it's in the debug output) and I don't see that happening when I'm logging in via ssh.

compinit is set to run here:
zinit ice wait'0' lucid atinit"ZINIT[COMPINIT_OPTS]=-C; zpcompinit; zpcdreplay"
zinit light zdharma/fast-syntax-highlighting

Does this help any?

**Comment by atwok** [comment]
the PZT::modules/utility messes with "make" and apparently this breaks make completions for me?  deleting that snippet and not loading it fixes my issues.

1)  why did this cause problems?
2)  why is compinit not running when I log in via ssh like it does when I do "rm ~/.zcopdump* && exec zsh -l"

thanks again for all your help

**Comment by psprint** [comment]
1. Apparently there's some code that removes the `utility` Prezto module from `$fpath`.
2. It rather has to be run. Could you check again?

**Comment by atwok** [comment]
I have some time to futz with this again.

So there is a directory called functions if you use PZT::modules/utility.  There's a forloop in the top level init.zsh for prezto the runs compinit on any "modules/*/functions" dirs if they exist.  It also adds it to the fpath (I think).  I'm not using that since I'm using zinit.  :)

I'm not doing anything to deal with this directory manually.  Is this something you want to handle in zinit since you already have code to pull in these snippets?  If not how should I handle this properly?

Thanks!

**Comment by atwok** [comment]
Do the colons in the path cause any issues?  I'm confused how the shell handles them since : is used to seperate entries in fpath and I dunno what else that zsh may care about.  :)

**Comment by atwok** [comment]
this piece here from prezto/init.zsh

```
 # Add functions to $fpath.
      fpath=(${pmodule_location}/functions(-/FN) $fpath)

      function {
        local pfunction

        # Extended globbing is needed for listing autoloadable function directories.
        setopt LOCAL_OPTIONS EXTENDED_GLOB

        # Load Prezto functions.
        for pfunction in ${pmodule_location}/functions/$~pfunction_glob; do
          autoload -Uz "$pfunction"
        done
      }
```

I'm guessing it's the autoload above that I don't have the equivalent of...

**Comment by psprint** [comment]
Yes, prezto autoloads the `functions/` directory contents and Zinit emulates this behavior. I would say that there's some code that removes the `PZT::…/functions` directory from `$fpath`. The colon does cause problems in `$FPATH` however when using `$fpath` then it works OK.
